engines/sci/graphics/cursor32.cpp

```diff
@@ -0,0 +1,394 @@
+		_cursor.rect = Common::Rect(_width, _height);
+		memset(_cursor.data, 255, _width * _height);
+		_cursor.skipColor = 255;
+
+		Buffer target(_width, _height, _cursor.data);
+		if (pixelDouble) {
+			view.draw(target, _cursor.rect, Common::Point(0, 0), false, 2, 2);
+		} else {
+			view.draw(target, _cursor.rect, Common::Point(0, 0), false);
+		}
+	} else {
+		_hotSpot = Common::Point(0, 0);
+		_width = _height = 1;
+		_cursor.data = (byte *)realloc(_cursor.data, _width * _height);
+		_cursor.rect = Common::Rect(_width, _height);
+		*_cursor.data = _cursor.skipColor;
+		_cursorBack.rect = _cursor.rect;
+		_cursorBack.rect.clip(_vmapRegion.rect);
+		if (!_cursorBack.rect.isEmpty()) {
+			readVideo(_cursorBack);
+		}
+	}
+
+	_cursorBack.data = (byte *)realloc(_cursorBack.data, _width * _height);
+	_drawBuff1.data = (byte *)realloc(_drawBuff1.data, _width * _height);
+	_drawBuff2.data = (byte *)realloc(_drawBuff2.data, _width * _height * 4);
+	_savedVmapRegion.data = (byte *)realloc(_savedVmapRegion.data, _width * _height);
+
+	unhide();
+}
+
+void GfxCursor32::readVideo(DrawRegion &target) {
+	if (g_sci->_gfxFrameout->_frameNowVisible) {
+		copy(target, _vmapRegion);
+	} else {
+		// NOTE: SSCI would read the background for the cursor directly out of
+		// video memory here, but as far as can be determined, this does not
+		// seem to actually be necessary for proper cursor rendering
+	}
+}
+
+void GfxCursor32::copy(DrawRegion &target, const DrawRegion &source) {
+	if (source.rect.isEmpty()) {
+		return;
+	}
+
+	Common::Rect drawRect(source.rect);
+	drawRect.clip(target.rect);
+	if (drawRect.isEmpty()) {
+		return;
+	}
+
+	const int16 sourceXOffset = drawRect.left - source.rect.left;
+	const int16 sourceYOffset = drawRect.top - source.rect.top;
+	const int16 drawWidth = drawRect.width();
+	const int16 drawHeight = drawRect.height();
+
+	byte *targetPixel = target.data + ((drawRect.top - target.rect.top) * target.rect.width()) + (drawRect.left - target.rect.left);
+	const byte *sourcePixel = source.data + (sourceYOffset * source.rect.width()) + sourceXOffset;
+
+	const int16 sourceStride = source.rect.width();
+	const int16 targetStride = target.rect.width();
+
+	for (int y = 0; y < drawHeight; ++y) {
+		memcpy(targetPixel, sourcePixel, drawWidth);
+		targetPixel += targetStride;
+		sourcePixel += sourceStride;
+	}
+}
+
+void GfxCursor32::setPosition(const Common::Point &position) {
+	const int16 scriptWidth = g_sci->_gfxFrameout->getCurrentBuffer().scriptWidth;
+	const int16 scriptHeight = g_sci->_gfxFrameout->getCurrentBuffer().scriptHeight;
+	const int16 screenWidth = g_sci->_gfxFrameout->getCurrentBuffer().screenWidth;
+	const int16 screenHeight = g_sci->_gfxFrameout->getCurrentBuffer().screenHeight;
+
+	_position.x = (position.x * Ratio(screenWidth, scriptWidth)).toInt();
+	_position.y = (position.y * Ratio(screenHeight, scriptHeight)).toInt();
+
+	g_system->warpMouse(_position.x, _position.y);
+}
```

Comment:
I wanted to make a note about that, so that these situations are known and that it's also known that there already is existing code to circumvent them.